### PR TITLE
Making scrypt more dynamic

### PR DESCRIPTION
**Pull request by awnumar**
A couple of changes to make scrypt cost-factors more dynamic and even allow user configuration.

### CODE REVIEW

File: crypto/crypto.go [hunk 1]
```diff
@@ -77,14 +77,14 @@ func Decrypt(base64EncodedCiphertext string, key []byte) []byte {
 }
 
 // DeriveKey derives a 32 byte encryption key from a password and identifier.
-func DeriveKey(password, identifier []byte) []byte {
-	derivedKey, _ := scrypt.Key(password, identifier, 1<<18, 8, 1, 32)
+func DeriveKey(password, identifier []byte, N uint) []byte {
```
Comment (tobbbles) [inline crypto/crypto.go]: argument names should be lowercase

**Review by tobbbles — DISMISSED** [review]

File: crypto/crypto.go (hunk 1 above)
Comment (awnumar) [reply crypto/crypto.go]: Not that argument. Scrypt's r and p can be lowercase, but N is capitalised.

File: crypto/crypto.go [hunk 2]
```diff
@@ -77,15 +77,15 @@ func Decrypt(base64EncodedCiphertext string, key []byte) []byte {
 }
 
 // DeriveKey derives a 32 byte encryption key from a password and identifier.
-func DeriveKey(password, identifier []byte) []byte {
-	derivedKey, _ := scrypt.Key(password, identifier, 1<<18, 8, 1, 32)
+func DeriveKey(password, identifier []byte, costFactor map[string]int) []byte {
+	derivedKey, _ := scrypt.Key(password, identifier, 1<<uint(costFactor["N"]), costFactor["r"], costFactor["p"], 32)
```
Comment (tobbbles) [inline crypto/crypto.go]: If costFactor[X] isn't set it'll default to the null value of int (zero) - do you want to add checks here to ensure the values are present?

Comment (awnumar) [reply crypto/crypto.go]: 🤔 Well there's a default set in `pocket.go` so the only way the value would be zero would be if the user manually and explicitly sets it to zero. Not sure if I should disallow that. Scrypt would throw an error and quit anyways so is it really needed?

Comment (tobbbles) [reply crypto/crypto.go]: Validating user inputs is always a good thing - I doubt the scrypt error is very user friendly, either

File: auxiliary/auxiliary.go [hunk 3]
```diff
@@ -232,13 +244,13 @@ func SaveSecrets(secrets map[string]interface{}) {
 	}
 
 	// Write the JSON to the disk.
-	ioutil.WriteFile("./.pocket/secrets", []byte(jsonFormattedSecrets), 0700)
+	ioutil.WriteFile("/home/awn/.pocket/secrets", []byte(jsonFormattedSecrets), 0700)
```
Comment (tobbbles) [inline auxiliary/auxiliary.go]: Change this.

Import `os/user` to get `u := user.Current()` and then use `u.HomeDir`. See https://godoc.org/os/user#User.

File: auxiliary/auxiliary_test.go [hunk 4]
```diff
@@ -43,20 +43,56 @@ func TestParseArgs(t *testing.T) {
...
-	if err.Error() != "[!] Invalid arguments" {
+	mode, _, err = ParseArgs(args)
+	if err.Error() != "[!] Invalid option" {
 		t.Error("Expected error; got ", err)
 	}
 	if mode != "" {
 		t.Error("Expected empty mode; got", mode)
 	}
+
+	args = []string{"./pocket", "get", "-c"}
```
Comment (tobbbles) [inline auxiliary/auxiliary_test.go:63]: Should should remove the `./` notation, as this wouldn't be present if the binary was in a users path

**Review by tobbbles — CHANGES_REQUESTED** [review]

File: auxiliary/auxiliary.go (hunk 3 above)
Comment (awnumar) [reply auxiliary/auxiliary.go]: We already do this.

File: auxiliary/auxiliary_test.go (hunk 4 above)
Comment (awnumar) [reply auxiliary/auxiliary_test.go:63]: It's being passed to a function that takes a slice of arguments, so it's not actually running anything, just emulating it being called with certain arguments.

**Pull request merged**